src/commonMain/kotlin/exchange.dydx.abacus/calculator/AdjustIsolatedMarginInputCalculator.kt

```diff
@@ -0,0 +1,141 @@
+import exchange.dydx.abacus.utils.mutable
+import exchange.dydx.abacus.utils.safeSet
+
+@Suppress("UNCHECKED_CAST")
+internal class AdjustIsolatedMarginInputCalculator(val parser: ParserProtocol) {
+    private val subaccountTransformer = SubaccountTransformer()
+
+    internal fun calculate(
+        state: Map<String, Any>,
+        parentSubaccountNumber: Int?,
+    ): Map<String, Any> {
+        val wallet = parser.asNativeMap(state["wallet"])
+        val isolatedMarginAdjustment = parser.asNativeMap(state["adjustIsolatedMargin"])
+        val childSubaccountNumber = parser.asInt(isolatedMarginAdjustment?.get("childSubaccountNumber"))
+        val type = parser.asString(isolatedMarginAdjustment?.get("type"))?.let {
+            IsolatedMarginAdjustmentType.invoke(it)
+        } ?: IsolatedMarginAdjustmentType.Add
+
+        return if (wallet != null && isolatedMarginAdjustment != null && type != null) {
+            val modified = state.mutable()
+            val parentTransferDelta = getModifiedTransferDelta(isolatedMarginAdjustment, true)
+            val childTransferDelta = getModifiedTransferDelta(isolatedMarginAdjustment, false)
+
+            val walletPostParentSubaccountTransfer =
+                subaccountTransformer.applyIsolatedMarginAdjustmentToWallet(
+                    wallet,
+                    subaccountNumber = parentSubaccountNumber,
+                    parentTransferDelta,
+                    parser,
+                    "postOrder",
+                )
+
+            val walletPostChildSubaccountTransfer =
+                subaccountTransformer.applyIsolatedMarginAdjustmentToWallet(
+                    wallet = walletPostParentSubaccountTransfer,
+                    subaccountNumber = childSubaccountNumber,
+                    childTransferDelta,
+                    parser,
+                    "postOrder",
+                )
+
+            val modifiedParentSubaccount = parser.asNativeMap(parser.value(walletPostChildSubaccountTransfer, "accounts.subaccounts.$parentSubaccountNumber"))
+            val modifiedChildSubaccount = parser.asNativeMap(parser.value(walletPostChildSubaccountTransfer, "accounts.subaccounts.$childSubaccountNumber"))
+            val modifiedIsolatedMarginAdjustment = finalize(isolatedMarginAdjustment, modifiedParentSubaccount, modifiedChildSubaccount, type)
+
+            modified["adjustIsolatedMargin"] = modifiedIsolatedMarginAdjustment
+            modified["wallet"] = walletPostChildSubaccountTransfer
+            modified
+        } else {
+            state
+        }
+    }
+
+    private fun getModifiedTransferDelta(
+        isolatedMarginAdjustment: Map<String, Any>,
+        isParentSubaccount: Boolean,
+    ): Map<String, Double>? {
+        val type = parser.asString(isolatedMarginAdjustment["type"])
+        val amount = parser.asDouble(isolatedMarginAdjustment["amount"])
+        if (type != null) {
+            when (type) {
+                "ADD" -> {
```

Comment:
[q] does kotlin support enums? would be nice to have switch cases not be string literals